### PR TITLE
Text exercise support

### PR DESCRIPTION
**Pull request by terlan98**
This PR adds support for assessing text exercises

The following features, in particular, are added for text exercises:
- **Starting** a new assessment
- **Creating inline feedbacks** with a word-level granularity
- **Colored highlights** based on feedback credits
- Creating **general feedbacks**
- **Deleting and editing** feedbacks
- **Canceling** an assessment
- **Saving** an assessment
- **Submitting** an assessment
- The aforementioned operations for text exercises in **exams**

There are also minor changes to the code pertaining to programming exercise assessments. Please also test if this PR breaks some existing functionality.

What is NOT included in this PR:
- Athene integration
- Search for submissions & read-only view

### CODE REVIEW

File: CodeEditor/Sources/CodeEditor/HighlightedRange.swift [hunk 1]
```diff
@@ -15,8 +15,6 @@ public final class HighlightedRange {
     public let range: NSRange
     /// Color to highlight the text with.
     public let color: UIColor
-    /// Corner radius of the highlight.
-    public let cornerRadius: CGFloat
```
Comment (pal03377) [inline CodeEditor/Sources/CodeEditor/HighlightedRange.swift:19]: ~Why is this removed? Was it not used? 🤔~
Edit: I just saw that the corner radius now is height*20% 👍

File: CodeEditor/Sources/CodeEditor/UXCodeTextView.swift [hunk 2]
```diff
@@ -98,6 +103,7 @@ final class UXCodeTextView: UXTextView, HighlightDelegate, UIScrollViewDelegate
             hlTextStorage.highlightDelegate = self
         }
         textContainer.exclusionPaths.append(UIBezierPath(rect: CGRect(x: 0, y: 0, width: Int(numViewWidth()) + 10, height: Int(INT_MAX))))
+        linkTextAttributes = [.foregroundColor: UIColor.black]
```
Comment (pal03377) [inline CodeEditor/Sources/CodeEditor/UXCodeTextView.swift]: Why do we need to set this? Can text submissions include links? Also, can the hardcoded value to `black` cause issues in dark mode?

File: CodeEditor/Sources/CodeEditor/UXCodeTextView.swift [hunk 3]
```diff
@@ -473,6 +492,42 @@ final class UXCodeTextView: UXTextView, HighlightDelegate, UIScrollViewDelegate
...
+        return textRangeToIntRange(selectedRange)
+    }
+    
+    /// Validates the `dragSelection` value
+    private func isSelectionValid() -> Bool {
+        guard !canSelectionIncludeHighlightedRanges else { return true }
+        
+        let highlightedIntRanges = highlightedRanges.compactMap({ Range($0.range) })
+        for range in highlightedIntRanges {
+            if self.dragSelection?.overlaps(range) == true {
```
Comment (pal03377) [inline CodeEditor/Sources/CodeEditor/UXCodeTextView.swift:529]:
```suggestion
            if self.dragSelection?.overlaps(range) {
```

File: Themis/Extensions/SubmissionExtension.swift [hunk 4]
```diff
@@ -10,6 +10,10 @@ import Foundation
 import SharedModels
 
 extension BaseSubmission {
+    func get<T>(as: T.Type = BaseParticipation.self) -> T? {
+        self as? T
```
Comment (pal03377) [inline Themis/Extensions/SubmissionExtension.swift]: Why can't you just so the `as? TextSubmission` (for example) directly when accessing the object?

File: Themis/Models/Feedback/ProgrammingFeedbackDetail.swift [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+import Foundation
+import SharedModels
+
+public struct ProgrammingFeedbackDetail: FeedbackDetail {
+    var file: Node?
+    var lines: NSRange?
+    var columns: NSRange?
+    
+    public mutating func buildArtemisFeedback(feedback baseFeedback: inout Feedback) {
```
Comment (pal03377) [inline Themis/Models/Feedback/ProgrammingFeedbackDetail.swift:16]: I think some documentation on how and when this is used would be nice here

File: Themis/Services/Assessment/AssessmentService.swift [hunk 6]
```diff
@@ -11,13 +11,27 @@ import Common
 
 protocol AssessmentService {
     /// Delete all saved feedback and release the lock of the submission
-    func cancelAssessment(submissionId: Int) async throws
+    func cancelAssessment(participationId: Int?, submissionId: Int) async throws
```
Comment (pal03377) [inline Themis/Services/Assessment/AssessmentService.swift:14]: I guess the participation ID is necessary for the text assessment API?

File: Themis/Services/Assessment/TextAssessmentServiceImpl.swift [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+class TextAssessmentServiceImpl: AssessmentService {
+    
+    let client = APIClient()
+    
+    // MARK: - Cancel Assessment
+    private struct CancelAssessmentRequest: APIRequest {
+        typealias Response = RawResponse
+        
+        var participationId: Int
+        var submissionId: Int
```
Comment (pal03377) [inline Themis/Services/Assessment/TextAssessmentServiceImpl.swift]: Can these be `let` variables?

File: Themis/Services/Assessment/TextAssessmentServiceImpl.swift [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        _ = try await client.sendRequest(CancelAssessmentRequest(participationId: participationId, submissionId: submissionId)).get()
+    }
+    
+    // MARK: - Save Assessment
+    private struct SaveAssessmentRequest: APIRequest {
+        typealias Response = RawResponse
+        
+        var participationId: Int
+        var resultId: Int
+        var assessmentDTO: TextAssessmentResult
```
Comment (pal03377) [inline Themis/Services/Assessment/TextAssessmentServiceImpl.swift]: Could these be `let` variables?

File: Themis/Services/Assessment/TextAssessmentServiceImpl.swift [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            .get()
+    }
+    
+    // MARK: - Submit Assessment
+    private struct SubmitAssessmentRequest: APIRequest {
+        typealias Response = RawResponse
+        
+        var participationId: Int
+        var resultId: Int
+        var assessmentDTO: TextAssessmentResult
```
Comment (pal03377) [inline Themis/Services/Assessment/TextAssessmentServiceImpl.swift]: `let` variables?

File: Themis/Services/Assessment/TextAssessmentServiceImpl.swift [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                                                 assessmentDTO: newAssessment))
+            .get()
+    }
+    
+    // MARK: - Fetch Participation For Submission
+    private struct GetParticipationRequest: APIRequest {
+        typealias Response = Participation
+        
+        var participationId: Int
+        var submissionId: Int
```
Comment (pal03377) [inline Themis/Services/Assessment/TextAssessmentServiceImpl.swift]: `let`?

File: Themis/Services/Submission/TextSubmissionServiceImpl.swift [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            "api/exercises/\(exerciseId)/text-submissions"
+        }
+    }
+    
+    func getAllSubmissions(exerciseId: Int) async throws -> [Submission] {
+        try await client.sendRequest(GetAllSubmissionsRequest(exerciseId: exerciseId)).get().0
+    }
+    
+    // MARK: - Get Tutor Submissions
+    private struct GetTutorSubmissionsRequest: APIRequest {
```
Comment (pal03377) [inline Themis/Services/Submission/TextSubmissionServiceImpl.swift:40]: "Tutor submissions" are submissions which have been graded by the tutor in the past, right? Maybe we could have a comment explaining this somewhere?

File: Themis/Services/Submission/UnknownSubmissionServiceImpl.swift [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+import Foundation
+import SharedModels
+import Common
+
+class UnknownSubmissionServiceImpl: SubmissionService {
+    
+    typealias SubmissionType = UnknownSubmission
+    
+    let defaultError = UserFacingError(title: "Exercise type not supported")
```
Comment (pal03377) [inline Themis/Services/Submission/UnknownSubmissionServiceImpl.swift]: The variable could be private

```suggestion
    private let defaultError = UserFacingError(title: "Exercise type not supported")
```

File: Themis/Supporting/ThemisUndoManager.swift [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+//  ThemisUndoManager.swift
+//  Themis
+//
+//  Created by Tarlan Ismayilsoy on 24.05.23.
+//
+
+import Foundation
+
+/// A wrapper class intended to perform UndoManager actions and publish notifications to update views
+class ThemisUndoManager {
```
Comment (pal03377) [inline Themis/Supporting/ThemisUndoManager.swift:11]: Is there a specific reason why you chose delegation over inheritance here? (I'm happy with it, just curious)

File: Themis/ViewModels/Assessment/AssessmentViewModel.swift [hunk 14]
```diff
@@ -8,37 +8,132 @@ class AssessmentViewModel: ObservableObject {
...
+    @MainActor
+    func initSubmission() async {
+        guard submission == nil else {
+            return
+        }
+        
+        switch exercise {
+        case .programming(exercise: _):
+            if readOnly {
+                if let participationId {
```
Comment (pal03377) [inline Themis/ViewModels/Assessment/AssessmentViewModel.swift:74]: What happens for `readOnly = true, participationId = nil`? Should it throw?

File: Themis/ViewModels/Assessment/AssessmentViewModel.swift [hunk 15]
```diff
@@ -8,37 +8,132 @@ class AssessmentViewModel: ObservableObject {
...
+            } else {
+                if let submissionId {
+                    await getSubmission(submissionId: submissionId)
+                } else {
+                    await initRandomSubmission()
+                }
+            }
+        case .text(exercise: _):
+            if readOnly {
+                // TODO: figure out which endpoint could be used instead
```
Comment (pal03377) [inline Themis/ViewModels/Assessment/AssessmentViewModel.swift:100]: Maybe throw for now?

File: Themis/ViewModels/Assessment/AssessmentViewModel.swift [hunk 16]
```diff
@@ -8,37 +8,132 @@ class AssessmentViewModel: ObservableObject {
...
+                                                                                                   submissionId: submissionId).baseParticipation
+            self.submission = fetchedParticipation.submissions?.last?.baseSubmission
+            self.participation = fetchedParticipation
+            assessmentResult.setComputedFeedbacks(basedOn: participation?.results?.last?.feedbacks ?? [])
+            assessmentResult.setReferenceData(basedOn: submission)
+            ThemisUndoManager.shared.removeAllActions()
+        } catch {
+            self.submission = nil
+            self.error = error
+            log.info(String(describing: error))
```
Comment (pal03377) [inline Themis/ViewModels/Assessment/AssessmentViewModel.swift]: Why not error log here?

```suggestion
            log.error(String(describing: error))
```

File: Themis/ViewModels/Assessment/AssessmentViewModel.swift [hunk 17]
```diff
@@ -47,23 +142,46 @@ class AssessmentViewModel: ObservableObject {
     }
 
     @MainActor
-    func getSubmission(id: Int) async {
+    func getSubmission(submissionId: Int) async {
+        guard !readOnly else {
```
Comment (pal03377) [inline Themis/ViewModels/Assessment/AssessmentViewModel.swift:167]: Will this ever be called if this assumption is incorrect? I would feel like it shouldn't be => it should throw because the expectation is violated

File: Themis/ViewModels/Assessment/AssessmentViewModel.swift [hunk 18]
```diff
@@ -47,23 +142,46 @@ class AssessmentViewModel: ObservableObject {
...
+            ThemisUndoManager.shared.removeAllActions()
+        } catch {
+            self.error = error
+            log.error(String(describing: error))
+        }
+    }
+    
+    @MainActor
+    func getReadOnlySubmission(participationId: Int) async {
+        guard readOnly else {
```
Comment (pal03377) [inline Themis/ViewModels/Assessment/AssessmentViewModel.swift:190]: Same as for `getSubmission`

File: Themis/ViewModels/Assessment/CodeEditorViewModel.swift [hunk 19]
```diff
@@ -281,3 +290,35 @@ class CodeEditorViewModel: ObservableObject {
...
+              let participationId,
+              let templateParticipationId else {
+            return
+        }
+        
+        withAnimation {
+            openFile(file: file, participationId: participationId, templateParticipationId: templateParticipationId)
+        }
+        
+        scrollUtils.range = inlineHighlights[file.path]?.first { $0.id == "\(feedback.id)" }?.range
```
Comment (pal03377) [inline Themis/ViewModels/Assessment/CodeEditorViewModel.swift]: Why is one ID a string and the other is not? 🤔

File: Themis/ViewModels/Assessment/TextExerciseRendererViewModel.swift [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+    private var submissionId: Int?
+    
+    
+    /// Sets this VM up based on the given participation and optional submission
+    /// - Parameters:
+    ///   - participation:
+    ///   - submission: Optional. Only used if the participation parameter does not contain a submission (this occurs when starting a new random assessment, for example)
+    @MainActor
+    func setup(basedOn participation: BaseParticipation?, and submission: BaseSubmission? = nil) {
+        var textSubmission = participation?.submissions?.last?.baseSubmission.get(as: TextSubmission.self)
```
Comment (pal03377) [inline Themis/ViewModels/Assessment/TextExerciseRendererViewModel.swift]:
```suggestion
        var textSubmission = participation?.submissions?.last?.baseSubmission as? TextSubmission
```

File: Themis/ViewModels/Assessment/TextExerciseRendererViewModel.swift [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+            
+        if textSubmission == nil { // no submission found in participation
+            textSubmission = submission?.get(as: TextSubmission.self)
```
Comment (pal03377) [inline Themis/ViewModels/Assessment/TextExerciseRendererViewModel.swift]:
```suggestion
            textSubmission = submission as? TextSubmission
```

File: Themis/ViewModels/Assessment/TextExerciseRendererViewModel.swift [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+        setupHighlights(basedOn: blocks, and: feedbacks)
+    }
+    
+    private func setupHighlights(basedOn blocks: [TextBlock], and feedbacks: [Feedback], shouldWipeUndo: Bool = true) {
+        blocks.forEach { block in
+            guard let blockId = block.id,
+                  let startIndex = block.startIndex,
+                  let endIndex = block.endIndex,
+                  let feedback = feedbacks.first(where: { $0.reference == block.id }),
+                  startIndex < endIndex else {
```
Comment (pal03377) [inline Themis/ViewModels/Assessment/TextExerciseRendererViewModel.swift]:
```suggestion
                  startIndex < endIndex
              else {
```

I don't know what Swiftlint thinks about this suggestion though

File: Themis/Views/Assessment/AssessmentView.swift [hunk 23]
```diff
@@ -1,212 +1,165 @@
...
+                                      submissionId: submissionId)
+        case .text(exercise: _):
+            TextAssessmentView(assessmentVM: assessmentVM,
+                               assessmentResult: assessmentResult,
+                               exercise: exercise,
+                               submissionId: submissionId,
+                               participationId: participationId,
+                               resultId: resultId)
+        default:
+            Text("Exercise not supported")
```
Comment (pal03377) [inline Themis/Views/Assessment/AssessmentView.swift:165]: I like this split!

File: Themis/Views/Assessment/CodeEditor/CodeView.swift [hunk 24]
```diff
@@ -59,7 +59,6 @@ struct CodeView: View {
     var editorFlags: CodeEditor.Flags {
         var flags: CodeEditor.Flags = []
         flags.insert(.selectable)
-        if colorScheme == .dark { flags.insert(.blackBackground) }
```
Comment (pal03377) [inline Themis/Views/Assessment/CodeEditor/CodeView.swift:74]: Why is this no longer necessary? I think I remember we added something like this because the background of the theme we chose was too grey for our taste. Has something changed there?

File: Themis/Views/Assessment/CorrectionSidebar/CorrectionSidebarView.swift [hunk 25]
```diff
@@ -30,24 +30,14 @@ struct CorrectionSidebarView: View {
...
-    
     var body: some View {
         VStack {
             sideBarElementPicker
             
             if !assessmentVM.loading {
                 ZStack {
                     ScrollView {
-                        ProblemStatementView(courseId: courseId, exerciseId: exercise?.id ?? -1)
+                        ProblemStatementView(courseId: courseVM.shownCourseID ?? -1, exerciseId: exercise?.id ?? -1)
```
Comment (pal03377) [inline Themis/Views/Assessment/CorrectionSidebar/CorrectionSidebarView.swift]: Suggestion: Maybe have an `if let` above and then show some kind of short error text if the `shownCourseID` or the `exercise.id` is `nil` respectively

File: Themis/Views/Assessment/CorrectionSidebar/FeedbackListView.swift [hunk 26]
```diff
@@ -114,28 +114,28 @@ struct FeedbackListView: View {
...
-                    codeEditorVM.deleteInlineHighlight(feedback: $0)
+                    feedbackDelegate?.onFeedbackDeletion($0)
                 }
             }
     }
 }
 
  struct FeedbackListView_Previews: PreviewProvider {
-    static let assessment = AssessmentViewModel(readOnly: false)
+     static let assessment = AssessmentViewModel(exercise: Exercise.mockText, readOnly: false)
```
Comment (pal03377) [inline Themis/Views/Assessment/CorrectionSidebar/FeedbackListView.swift]:
```suggestion
    static let assessment = AssessmentViewModel(exercise: Exercise.mockText, readOnly: false)
```
One space too much?

File: Themis/Views/Assessment/Pane/RightGripView.swift [hunk 27]
```diff
@@ -33,6 +33,7 @@ struct RightGripView: View {
             if paneVM.dragWidthRight > 0 {
                 Color.themisPrimary
                     .frame(maxWidth: 7, maxHeight: .infinity)
+                    .ignoresSafeArea()
```
Comment (pal03377) [inline Themis/Views/Assessment/Pane/RightGripView.swift:36]: Did the safe area cause issues with the grip? I wouldn't have thought of this!

File: Themis/Views/Assessment/Text Exercise/TextExerciseRenderer.swift [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+struct TextExerciseRenderer_Previews: PreviewProvider {
+    @State static var textVM = TextExerciseRendererViewModel()
+    
+    static var previews: some View {
+        TextExerciseRenderer(textExerciseRendererVM: textVM)
+            .previewInterfaceOrientation(.landscapeRight)
+            .onAppear {
+//                textVM.setup(basedOn: Submission.mockText.baseSubmission.participation)
```
Comment (pal03377) [inline Themis/Views/Assessment/Text Exercise/TextExerciseRenderer.swift]: Why is this commented?

File: Themis/Views/Assessment/TextAssessmentView.swift [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+        }
+        .onReceive(didStartNextAssessment, perform: { _ in
+            guard assessmentVM.submission != nil && assessmentVM.participation != nil else {
+                return
+            }
+            textExerciseRendererVM.setup(basedOn: assessmentVM.participation, and: assessmentVM.submission)
+            ensureResultId(force: true)
+        })
+        .onAppear {
+            assessmentVM.fontSize = 19.0
```
Comment (pal03377) [inline Themis/Views/Assessment/TextAssessmentView.swift:50]: That's an oddly specific font size number 😅  Why not 18?

**Review by pal03377 — COMMENTED** [review]
I did a code review pass, I essentially just have some questions and suggestions. I didn't get to test it yet because after @terlan98 helped me fix my XCode build in a meeting today, we found a bug which I'm not waiting to be resolved before testing more.

File: CodeEditor/Sources/CodeEditor/UXCodeTextView.swift (hunk 2 above)
Comment (terlan98) [reply CodeEditor/Sources/CodeEditor/UXCodeTextView.swift]: Apparently, highlights are implemented as links (they contain the id of the feedback used for editing) and this causes the text inside highlights to be blue. This line overrides the default link color. 

It looked fine in dark mode but now I changed it to `UIColor.label`, so it looks even better :)

File: CodeEditor/Sources/CodeEditor/UXCodeTextView.swift (hunk 3 above)
Comment (terlan98) [reply CodeEditor/Sources/CodeEditor/UXCodeTextView.swift:529]: `self.dragSelection?.overlaps(range)` is of type `Bool?`. So this suggestion does not work, unfortunately.

File: Themis/Views/Assessment/CodeEditor/CodeView.swift (hunk 24 above)
Comment (terlan98) [reply Themis/Views/Assessment/CodeEditor/CodeView.swift:74]: Because I have added a new dark-mode compatible color to the Assets and used it as follows:
https://github.com/ls1intum/Themis/blob/ddb28f19d48c4d7fce10c824882bac0f2bc8e61f/CodeEditor/Sources/CodeEditor/UXCodeTextViewRepresentable.swift#L258

File: Themis/Views/Assessment/TextAssessmentView.swift (hunk 29 above)
Comment (terlan98) [reply Themis/Views/Assessment/TextAssessmentView.swift:50]: 😅 18 seemed too small

File: Themis/Views/Assessment/Pane/RightGripView.swift (hunk 27 above)
Comment (terlan98) [reply Themis/Views/Assessment/Pane/RightGripView.swift:36]: Yes, the vertical line was cut at the bottom of the screen due to the safe area 🙂

File: Themis/Services/Assessment/AssessmentService.swift (hunk 6 above)
Comment (terlan98) [reply Themis/Services/Assessment/AssessmentService.swift:14]: Yes. It may also be needed for other exercise types, such as modeling.

File: Themis/Supporting/ThemisUndoManager.swift (hunk 13 above)
Comment (terlan98) [reply Themis/Supporting/ThemisUndoManager.swift:11]: Inheritance, unfortunately, does not work since the `registerUndo(withTarget:handler:)` method is declared in an extension of `UndoManager`. This prevents overriding it in a subclass.

File: Themis/ViewModels/Assessment/AssessmentViewModel.swift (hunk 14 above)
Comment (terlan98) [reply Themis/ViewModels/Assessment/AssessmentViewModel.swift:74]: Good catch! I will set the error variable, which will show an alert to the user in this case.

**Comment by terlan98** [comment]
Thanks a lot for your review @pal03377 🙏 
I've implemented new changes and also fixed the bug we found during our last meeting.

File: CodeEditor/Sources/CodeEditor/UXCodeTextViewRepresentable.swift [hunk 30]
```diff
@@ -152,8 +152,10 @@ public struct UXCodeTextViewRepresentable: UXViewRepresentable {
         
         public func textView(_ textView: UITextView, shouldInteractWith URL: URL,
                              in characterRange: NSRange, interaction: UITextItemInteraction) -> Bool {
-            if interaction == .invokeDefaultAction {
-                self.parent.editorBindings.feedbackForSelectionId.wrappedValue = URL.absoluteString
+            if interaction == .invokeDefaultAction,
+               let feedbackId = UUID(uuidString: URL.absoluteString)
```
Comment (pal03377) [inline CodeEditor/Sources/CodeEditor/UXCodeTextViewRepresentable.swift:156]: Why is this declaration in the `if` condition part? Can't it be in the block?

**Review by pal03377 — APPROVED** [review]
The code changes since my last review look good 👍 Also, I could now test text exercise assessments in the Themis app (simulator) 🎉 Well done!

I noticed that there is no context menu option for "Feedback" in text exercises. Is this intended?

Text exercise:
screenshot-2023-06-28_001474

Programming exercise:
screenshot-2023-06-28_001475

File: CodeEditor/Sources/CodeEditor/UXCodeTextViewRepresentable.swift (hunk 30 above)
Comment (terlan98) [reply CodeEditor/Sources/CodeEditor/UXCodeTextViewRepresentable.swift:156]: `UUID(uuidString:)` returns an optional. It should normally never be nil, since all highlights are linked to a feedback id. However, if it somehow does become nil, the block should not be executed. That's why it's a part of the if condition.

**Comment by terlan98** [comment]
> The code changes since my last review look good 👍 Also, I could now test text exercise assessments in the Themis app (simulator) 🎉 Well done!
> 
> I noticed that there is no context menu option for "Feedback" in text exercises. Is this intended?
> 
> Text exercise: screenshot-2023-06-28_001474
> 
> Programming exercise: screenshot-2023-06-28_001475

Yes, this is intended. Text assessments allow only word-level granularity for inline feedback. The context menu, however, allows adding feedback at character-level granularity as well. That's why I had to disable it.

**Comment by maximiliansoelch** [comment]
When testing, I found an issue while viewing the feedback I gave via Themis in Artemis.

This is how the feedback looks in Themis:
Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-07-10 at 18 38 10

However, when I open it in Artemis, the feedback is wrong, and some of the original submission text is missing. When I delete the inline feedback, the correct submission is shown again:

https://github.com/ls1intum/Themis/assets/9954674/3fa70923-b9c3-4234-8139-04614db14e52

**Comment by terlan98** [comment]
> When testing, I found an issue while viewing the feedback I gave via Themis in Artemis.
> 
> This is how the feedback looks in Themis: Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-07-10 at 18 38 10
> 
> However, when I open it in Artemis, the feedback is wrong, and some of the original submission text is missing. When I delete the inline feedback, the correct submission is shown again:
> 
>  Screen.Recording.2023-07-10.at.18.37.43.mov

Thanks for the review 🙂 There's currently an issue with overlapping text blocks on Artemis. I made some changes on my latest commit that will reduce the likelihood of this happening, but the issue itself is on Artemis, and it can be reproduced even without using the Themis app. Please see the following issue: https://github.com/ls1intum/Artemis/issues/6903 

I've already created an [Artemis PR](https://github.com/ls1intum/Artemis/pull/6916) to fix the aforementioned issue.

**Review by maximiliansoelch — APPROVED** [review]
Tested locally, giving feedback on text exercises works as expected.
Code looks good 👍
Great work!
